appmap/_implementation/recorder.py

```diff
@@ -22,10 +23,14 @@ class Recorder(ABC):
     def events(self):
         return self._events
 
-    @abstractmethod
-    def next_event_id(self):
-        self._next_event_id += 1
-        return self._next_event_id
+    _next_event_id = 0
+    _next_event_id_lock = threading.Lock()
+
+    @classmethod
+    def next_event_id(cls):
+        with cls._next_event_id_lock:
+            cls._next_event_id += 1
+            return cls._next_event_id
```

Comment:
What was the interference of remote requests with remote recording?

Was part of it a race condition over event ids?  If event ids are unique per thread then why put them behind a lock?

Reply:
There were three two aspects to the interference:
* remote recording and per-request recording were both trying to use Recorder.get_current().enabled to manage the state of remote recording
* new event ids weren't getting issued properly

I fixed the first by tracking the state of remote recording with a `ContextVar` with a boolean value. I also simplified the code somewhat by separating it from the middleware that captures `http_server` events.

I fixed the second by having the next event id be global (it's a class variable of `Recorder`, which is why there's locking around updating it). Careful management of resetting it meant that no tests need to change. That management probably also means it's a bit fragile, but perhaps that an issue for another time.

One thing that might be surprising about this PR: there aren't any updates for the expected AppMaps. This is actually because of the problem identified with AppMap correctness: the only route in the test Flask app is implemented as a [function in module](https://github.com/getappmap/appmap-python/blob/bb33143582c68433d59800fd8e1f050bd4ba36f1/appmap/test/data/flask/app.py#L15) , so it doesn't show up in the recordings.

WDYT, does it seem worth adding a test that show enabling both kinds of recording results in a proper AppMap?

Reply:
I think
- this pr is worth shipping as is
- it's worth adding a low priority issue to revisit event ids.  If event ids are unique per thread then it appears possible to not need a global lock for event ids.  At this point this looks like an optimization, so it doesn't appear like a priority.

[test_record_request_appmap_enabled_requests_enabled_and_remote](https://github.com/getappmap/appmap-python/blob/master/appmap/test/test_django.py#L251) collects both kinds of recording results.  I wonder if at least one of the many threads there contains data collected for remote recording.  Maybe all it needs is an assert or two in [record_request_thread](https://github.com/getappmap/appmap-python/blob/master/appmap/test/web_framework.py#L371).

Reply:
This doesn't work for Django for me:
```
APPMAP=true APPMAP_RECORD_REQUESTS=true APPMAP_OUTPUT_DIR=/tmp python3 manage.py runserver 0.0.0.0:8000

curl -s localhost:8000/_appmap/record
```
The `_appmap/record` doesn't exist.

Reply:
Try running it like this:
```
PYTHONPATH=init APPMAP=true APPMAP_RECORD_REQUESTS=true APPMAP_OUTPUT_DIR=/tmp  FLASK_DEBUG=1 FLASK_APP=app.py flask run -p 8000
```
Setting `PYTHONPATH` here simulates what happens when the package is really installed.